main/src/main/java/cgeo/geocaching/utils/OfflineTranslateUtils.java

```diff
@@ -127,21 +132,38 @@ public static void getTranslator(final Activity activity, final Language sourceL
     }
 
     public static void translateParagraph(final Translator translator, final OfflineTranslateUtils.Status status, final String text, final Consumer<String> consumer, final Consumer<Exception> errorConsumer) {
-        final List<String> origText = Arrays.asList(text.split("\n"));
-        final List<String> translatedText = new ArrayList<>();
-        for (String p : origText) {
-            translator.translate(p)
-                    .addOnSuccessListener(translation -> {
-                        translatedText.add(translation);
-                        if (origText.size() == translatedText.size()) {
-                            consumer.accept(String.join("\n", translatedText));
-                            status.updateProgress();
-                        }
-                    })
-                    .addOnFailureListener(e -> {
-                        status.abortTranslation();
-                        errorConsumer.accept(e);
-                    });
+        final Document document = Jsoup.parse(text);
+        final Elements elements = document.children().select("*");
+        final AtomicInteger remaining = new AtomicInteger(elements.size());
+        for (Element element : elements) {
+            final String current = element.ownText();
+            if (StringUtils.isBlank(current)) {
+                remaining.decrementAndGet();
+            } else {
+                // translate
+                final String[] origText = current.split("\n");
```

Comment:
When translating HTML text snippets it shouldn't be necessary to preserve line breaks. I think it should be enough to translate `current` in one go.

Reply:
Thanks, @eddiemuc, for taking the time to review.

The `split()` code is from the original parsing. As `ownText()` seems to normalize the text, removing newlines along the way, it's indeed no longer necessary. In JSoup docs I found a method `wholeOwnText()` which seems to preserve such whitespace, which might solve the problem with the vanishing paragraph breaks - I'll play with it (and either find a solution for the paragraphs or remove the special newline handling).

Reply:
> The `split()` code is from the original parsing. 

I believe the original parsing needed this because it parsed text already converted by `Html.fromHtml()`. In the result of this, `<p>` and `<br>` was already converted into \n

However, your approach parses HTML text nodes in which there should no `\n` being present or they shouldn't have any effect on the display anyway.

Reply:
> In JSoup docs I found a method wholeOwnText() which seems to preserve such whitespace...

I don't believe this solves the problem. White spaces/line breaks in HTML **should** be ignored; keeping them would be a bug.

Reply:
After adapting + comparing with the original HTML: Using `wholeOwnText()` does not make a difference here, as the line breaks in that specific case come from a couple of `<br>` between even more `<span>`, and JSoup somehow strips the `<br>` :-(

I have remove handling of newlines for now.

Reply:
> > In JSoup docs I found a method wholeOwnText() which seems to preserve such whitespace...
> 
> I don't believe this solves the problem. White spaces/line breaks in HTML **should** be ignored; keeping them would be a bug.

My assumption was, that `HTML.from()` somehow translates two consecutive newlines into a paragraph, but that assumption was wrong. And, having looked in the source for the specific cache, the "paragraphs" were done quite differently.